src/algebra/big_operators.lean

```diff
@@ -338,6 +350,30 @@ by haveI := classical.dec_eq α; exact
     by rw [prod_image]; exact assume x _ y _, subtype.eq
   ... = _ : by rw [attach_image_val]
 
+/-- A product over `s.subtype p` equals one over `s.filter p`. -/
+@[to_additive "A sum over `s.subtype p` equals one over `s.filter p`."]
+lemma prod_subtype_eq_prod_filter (f : α → β) {p : α → Prop} [decidable_pred p] :
+  ∏ x in s.subtype p, f x = ∏ x in s.filter p, f x :=
+begin
+  conv_lhs {
+    erw ←prod_map (s.subtype p) (function.embedding.subtype _) f
+  },
+  exact prod_congr (subtype_map _) (λ x hx, rfl)
+end
+
+/-- If all elements of a `finset` satisfy the predicate `p`, a product
+over `s.subtype p` equals that product over `s`. -/
+@[to_additive "If all elements of a `finset` satisfy the predicate `p`, a sum
+over `s.subtype p` equals that sum over `s`."]
+lemma prod_subtype_of_mem (f : α → β) {p : α → Prop} [decidable_pred p]
+    (h : ∀ x ∈ s, p x) : ∏ x in s.subtype p, f x = ∏ x in s, f x :=
```

Comment:
You've added this lemma, but then not used it. Are you still planning on refactoring the big proof further?

Reply:
Changing anything in `algebra.big_operators` or `data.finset` means Lean wants to recompile large parts of mathlib. So my workflow for adding lemmas there was to write those lemmas and their users elsewhere, then move them to the right place, then push those lemmas and go away and wait for CI to recompile mathlib before seeing if the code to use the lemmas still works with the final version of the lemmas in the correct place (and only then pushing the code using the lemmas if it does work). I've now pushed the refactor to use those lemmas (but still need to refactor the first half of the proof, which could involve adding more such lemmas).